[Serve]Make http retries tunable (#32532)
ray-project/ray: doc/source/serve/performance.md

By default, Serve lets client HTTP requests run to completion no matter how long they take. However, slow requests could bottleneck the replica processing, blocking other requests that are waiting. It's recommended that you set an end-to-end timeout, so slow requests can be terminated and retried at another replica.

You can set an end-to-end timeout for HTTP requests by setting the `RAY_SERVE_REQUEST_PROCESSING_TIMEOUT_S` environment variable. HTTP Proxies will wait for that many seconds before terminating an HTTP request and retrying it at another replica. This environment variable should be set on every node in your Ray cluster, and it cannot be updated during runtime.

(serve-performance-http-retry)=

> ### Set request retry times

Reviewer: ```suggestion
### Set request retry limit
```

> By default, the Serve HTTP proxy retries `10` times if the response is not received due to failures (e.g. network disconnect, request timeout etc.).

Reviewer 2: ```suggestion
By default, the Serve HTTP proxy retries up to `10` times when a response is not received due to failures (e.g. network disconnect, request timeout, etc.).
```

> By default, Serve HTTP proxy will retry `10` times if the response is not able to be back due to failures (e.g. network disconnect, request timeout etc.).

Reviewer 3: ```suggestion
By default, the Serve HTTP proxy retries `10` times if the response is not received due to failures (e.g., network disconnect, request timeout etc.)
```

Reviewer 3: Not sure if `Serve HTTP proxy` is a proper noun or a generic noun. Feel free to adjust.

Author: HTTP Proxy is commonly used in the ray serve doc. I will keep it as it is.

> You can set number of retries based on your requirements by setting the `RAY_SERVE_MAX_REPLICA_FAILURE_RETRIES` environment variable. HTTP proxy will at most retry that times before sending task errors response to the client.

Reviewer: We should have `HTTP` in this name given it only applies to HTTP proxy

Also, this is directly related to the request timeout, can you mention that here?

> You can set number of retries based on your requirements by setting the `RAY_SERVE_MAX_REPLICA_FAILURE_RETRIES` environment variable. HTTP proxy will at most retry that times before sending task errors response to the client.

Reviewer 3: ```suggestion
You can set number of retries based on your requirements by setting the `RAY_SERVE_MAX_REPLICA_FAILURE_RETRIES` environment variable. HTTP proxy retries retries that many times at most before sending a task error response to the client.
```

> You can set number of retries based on your requirements by setting the `RAY_SERVE_HTTP_MAX_REPLICA_FAILURE_RETRIES` environment variable. HTTP proxy retries that many times at most before sending a task error response to the client. The number of retries will affect e2e latency of your HTTP requests.

Reviewer 2: ```suggestion
You can set the number of retries based on your requirements by setting the `RAY_SERVE_HTTP_MAX_REPLICA_FAILURE_RETRIES` environment variable. The HTTP proxy will retry that many times at most before sending a task error response to the client. The number of retries will affect end-to-end latency of your HTTP requests.
```

Reviewer 2: > The number of retries will affect e2e latency of your HTTP requests.

Could you give more detail here? How will the latency be affected? I think we should mention that we do an exponential backoff here.

> You can set the number of retries based on your requirements by setting the `RAY_SERVE_HTTP_MAX_REPLICA_FAILURE_RETRIES` environment variable. The HTTP proxy will retry that many times at most before sending a task error response to the client. The number of retries will affect end-to-end latency of your HTTP requests. (Ray Serve will do expotential backoff retry).

Reviewer 2: ```suggestion
You can set the number of retries based on your requirements by setting the `RAY_SERVE_HTTP_MAX_REPLICA_FAILURE_RETRIES` environment variable. The HTTP proxy will retry that many times at most before sending a task error response to the client. The number of retries will affect the end-to-end latency of your HTTP requests since Ray Serve does exponential backoff for retries.
```

> You can set the number of retries based on your requirements by setting the `RAY_SERVE_HTTP_MAX_REPLICA_FAILURE_RETRIES` environment variable. The HTTP proxy will retry that many times at most before sending a task error response to the client. The number of retries will affect the end-to-end latency of your HTTP requests since Ray Serve does exponential backoff for retries.

Reviewer: ```suggestion
You can tune the number of retries by setting the `RAY_SERVE_HTTP_MAX_RETRIES` environment variable. The HTTP proxy will retry that many times at most before sending a task error response to the client. The number of retries will affect the end-to-end latency of your HTTP requests because Ray Serve does exponential backoff for retries.
```
suggest to remove `REPLICA_FAILURE` because this retry is not limited to replica failures, also happens if network disconnects, request times out, etc. (as specified above)

Author: good point, updated to RAY_SERVE_HTTP_REQUEST_MAX_RETRIES.
